R/colors.R

```diff
@@ -268,35 +280,58 @@ colorFactor <- function(palette, domain, levels = NULL, ordered = FALSE,
 NULL
 
 
-safePaletteFunc <- function(pal, na.color, alpha) {
-  toPaletteFunc(pal, alpha=alpha) %>% filterRGB() %>% filterZeroLength() %>%
-    filterNA(na.color) %>% filterRange()
+safePaletteFunc <- function(pal, na.color, alpha, nlevels = NULL) {
+  toPaletteFunc(pal, alpha=alpha, nlevels = nlevels) %>%
+    filterRGB() %>%
+    filterZeroLength() %>%
+    filterNA(na.color) %>%
+    filterRange()
 }
 
-toPaletteFunc <- function(pal, alpha) {
+# nlevels is a positive or negative integer (or integral number) indicating the
+# number of levels to use for a discrete scale (i.e. factor, i.e. qualitative,
+# i.e. categorical); or NULL if it is a continuous scale. A negative value means
+# that the user has asked for a "reversed" palette, so pull from the tail of the
+# color palette rather than from the head.
+#
+# (Previous versions of this code didn't have nlevels and simply interpolated
+# between colors in a qualitative palette--clearly the wrong thing to do.)
+toPaletteFunc <- function(pal, alpha, nlevels) {
   UseMethod("toPaletteFunc")
 }
 
 # Strings are interpreted as color names, unless length is 1 and it's the name
-# of an RColorBrewer palette
-toPaletteFunc.character <- function(pal, alpha) {
+# of an RColorBrewer palette that is marked as qualitative
+toPaletteFunc.character <- function(pal, alpha, nlevels) {
   if (length(pal) == 1 && pal %in% row.names(RColorBrewer::brewer.pal.info)) {
-    return(scales::colour_ramp(
-      RColorBrewer::brewer.pal(RColorBrewer::brewer.pal.info[pal, 'maxcolors'], pal),
-      alpha = alpha
-    ))
+    paletteInfo <- RColorBrewer::brewer.pal.info[pal,]
+    colors <- RColorBrewer::brewer.pal(paletteInfo$maxcolors, pal)
+
+    if (!is.null(nlevels) && paletteInfo$category == "qual") {
```

Comment:
I feel like you could drop `paletteInfo$category == "qual"` - ideally you should use the colours as given whenever there are enough already (ie. you might have an ordered factor)